### PR TITLE
Optimise the update logic of the frontend table to cope with duplicate service and workload addresses.

### LINKED ISSUE #1147: Optimising the frontend table for variable service addresses.
**What would you like to be added**:
Optimise the update logic of the frontend table to cope with variable service addresses.

**Why is this needed**:

Previously, the design only considered cluster service. But after the introduction of `serviceEntry`, the existing processing logic will have loopholes, resulting in data residuals. Therefore, the processing logic needs to be optimised to meet the processing requirements of the out-of-cluster service.

### PR DESCRIPTION
**Pull request by LiZhenCheng9527**
**What type of PR is this?**
/kind enhancement

**What this PR does / why we need it**:
The logic of the existing frontend map in handling service update does not consider the scenario of service address change. 
https://github.com/kmesh-net/kmesh/blob/7425245e40171db1759198cc9ced8d20cc0a5d66/pkg/controller/workload/workload_processor.go#L691-L695
When updating, the new service will be used directly in frontend map update, and will not process the service information that already exists in the map, which will result in the old service information being left. 

So, I add a function  can compare two slices and return the data added to a over b and the data missing from b over a.
And Optimising the frontend table for variable service addresses, to process the old data in the frontend map.

**Which issue(s) this PR fixes**:
Fixes #1147 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/controller/workload/workload_processor.go [hunk 1]
```diff
@@ -682,10 +702,24 @@ func (p *Processor) updateServiceMap(service *workloadapi.Service) error {
...
+		// Compare the addresses of the old and new maps to avoid residual.
+		// If the data can be found in the km_service map, it is also stored in the serviceCache.
+		newServiceAddress := utils.GetAddressesFromService(service)
+		oldServiceAddress := utils.GetAddressesFromService(oldService)
+		_, removeServiceAddress := utils.CompareIpByte(newServiceAddress, oldServiceAddress)
+
+		frontKey := bpf.FrontendKey{}
+		for _, address := range removeServiceAddress {
+			nets.CopyIpByteFromSlice(&frontKey.Ip, address)
+			if err := p.bpf.FrontendDelete(&frontKey); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/controller/workload/workload_processor.go:715]: we can abstract a new function as i see it is used by workload handler too

File: pkg/utils/slices_handler.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	for _, item := range aSet {
+		aNew = append(aNew, item)
+	}
+
+	return aNew, bMissing
+}
+
+func GetAddressesFromService(service *workloadapi.Service) [][]byte {
```
Comment (hzxuzhonghu) [inline pkg/utils/slices_handler.go:70]: This is not related to slice processing, we can move it to controller/workload pkg

Comment (LiZhenCheng9527) [reply pkg/utils/slices_handler.go:70]: I will move it to pkg/nets

File: pkg/controller/workload/workload_processor.go (hunk 1 above)
Comment (LiZhenCheng9527) [reply pkg/controller/workload/workload_processor.go:715]: Done

File: pkg/controller/workload/workload_processor.go [hunk 3]
```diff
@@ -486,6 +487,22 @@ func (p *Processor) handleWorkload(workload *workloadapi.Workload) error {
 		log.Errorf("updateWorkloadInFrontendMap %s failed: %v", workload.Uid, err)
 		return err
 	}
+	if oldWorkload != nil {
+		// To be able to find a workload in the workloadCache,
+		// you need to determine whether the address of the workload has changed or not.
+		// And clean up the residue
+		newWorkloadAddresses := workload.GetAddresses()
+		oldWorkloadAddresses := oldWorkload.GetAddresses()
+		_, removeAddresses := utils.CompareIpByte(newWorkloadAddresses, oldWorkloadAddresses)
```
[the remark this answers is not in the document]
Comment (LiZhenCheng9527) [reply pkg/controller/workload/workload_processor.go:496]: Done

File: pkg/utils/slices_handler.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/utils/slices_handler.go]: isn't it  related to service object, or you can provide a method to it, like `ResourceName`

File: pkg/controller/workload/workload_processor.go [hunk 4]
```diff
@@ -682,10 +702,19 @@ func (p *Processor) updateServiceMap(service *workloadapi.Service) error {
 				return nil
 			}
 		}
+
+		// Compare the addresses of the old and new maps to avoid residual.
+		// If the data can be found in the km_service map, it is also stored in the serviceCache.
+		newServiceAddress := getAddressesFromService(service)
+		oldServiceAddress := getAddressesFromService(oldService)
+		removeServiceAddress := nets.CompareIpByte(newServiceAddress, oldServiceAddress)
+		if err := p.deleteOutdataFrontend(removeServiceAddress); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/controller/workload/workload_processor.go:711]: donot call outdated, by this you make the function limited to process outdated address only. But actually it provide general functionality

File: pkg/controller/workload/workload_processor.go [hunk 5]
```diff
@@ -1022,3 +1052,30 @@ func (p *Processor) storeWorkloadPolicies(uid string, polices []string) {
 		log.Errorf("storeWorkloadPolicies failed, workload %s, err: %s", uid, err)
 	}
 }
+
+func (p *Processor) deleteOutdataFrontend(addresses [][]byte) error {
+	frontKey := bpf.FrontendKey{}
+	for _, address := range addresses {
+		nets.CopyIpByteFromSlice(&frontKey.Ip, address)
+		if err := p.bpf.FrontendDelete(&frontKey); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/controller/workload/workload_processor.go:1060]: In case it delete a service frontend map i think we should do some protection. There are some other work we can do, maybe you should do that in another pr of you

File: pkg/controller/workload/workload_processor.go (hunk 4 above)
Comment (LiZhenCheng9527) [reply pkg/controller/workload/workload_processor.go]: Done

File: pkg/nets/nets_test.go [hunk 6]
```diff
@@ -67,3 +67,40 @@ func TestCopyIpByteFromSlice(t *testing.T) {
 		})
 	}
 }
+
+func TestCompareIpByte(t *testing.T) {
```
Comment (hzxuzhonghu) [inline pkg/nets/nets_test.go:71]: Recommend table driven test

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm
